enhancements/network/network-policy-audit-logging.md

```diff
@@ -0,0 +1,357 @@
+ // not using OVN.
+ // +optional
+ HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
+ // ipsecConfig enables and configures IPsec for pods on the pod network within the
+ // cluster.
+ // +optional
+ IPsecConfig *IPsecConfig `json:"ipsecConfig,omitempty"`
+ 
+ <--- BEGIN NEW CODE --->
+ // aclLoggingRateLimit allows for the configuring of max acl logging rate
+ // defaults to 20 messages per second
+ // +optional
+ aclLoggingRateLimit  *uint32 `json:"aclLoggingRateLimit,omitempty"`
+ <--- END NEW CODE --->
+
+}
+```
+
+The flag `aclLoggingRateLimit` defaults to `20` messages per second.
+
+This will result in an updated OVNKubernetes Operator configuration object: `(.spec.defaultNetwork.ovnKubernetesConfig)` which will enable cluster-wide configuration of ACL logging at cluster installation time:
+
+```go
+spec:
+  defaultNetwork:
+    type: OVNKubernetes
+    ovnKubernetesConfig:
+      mtu: 1400
+      genevePort: 6081
+      ipsecConfig: {}
+      aclLoggingRateLimit: 20
+```
+
+To enable the ACL logging on a per namespace basis using metadata annotations in the namespace's definition like the following
+
+```go
+kind: Namespace
+apiVersion: v1
+metadata:
+  name: tenantA
+  annotations:
+    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "notice" }'
+```
+
+or
+
+```go
+kind: Namespace
+apiVersion: v1
+metadata:
+  name: tenantB
+  annotations:
+    k8s.ovn.org/acl-logging: '{ "deny": "notice" }'
+```
+
+The logging can be activated for either allow, drop, or allow and drop actions. The severity must be one of `alert`, `warning`,
+`notice`, `info`, or `debug` as described in [OVN documentation](http://www.openvswitch.org/support/dist-docs/ovn-nbctl.8.html).
+
+each level is listed in order or descending severity
+
+```go
+alert    A major failure forced a process to abort.
+
+warning  A high-level operation or a subsystem failed.  Attention
+         is warranted.
+
+notice   A low-level operation failed, but higher-level subsystems
+         may be able to recover.
+
+info     Information that may be useful in retrospect when
+         investigating a problem.
+
+debug    Information useful only to someone with intricate
+         knowledge of the system, or that would commonly cause too-
+         voluminous log output.  Log messages at this level are not
+         logged by default.
+```
+
+### User Stories
+
+#### Story 1
+
+As an Openshift user I want to see what Network Policies and accompanied ACLs are dropping the most traffic in a cluster. For example, when a packet runs into a `drop` ACL it will log a message similar to the following:
+
+ ```sh
+  2021-01-05T17:34:02.675Z|00004|acl_log(ovn_pinctrl0)|INFO|name="<Network Policy Name>", verdict=drop, severity=info: icmp,vlan_tci=0x0000,dl_src=50:54:00:00:00:02,dl_dst=50:54:00:00:00:01,nw_src=192.168.0.3,nw_dst=192.168.0.2,nw_tos=0,nw_ecn=0,nw_ttl=64,icmp_type=8,icmp_code=0
```

Comment:
For user to be able to make sense of this, the source/destination podname and namespace would need to also be included, otherwise they'll need to manually resolve the IP which is potentially valid for a short period of time, which IMO would limit the usefulness of this feature.

As a user I'd like to see something like: "NetworkPolicy namespace1/X is dropping icmp/8 traffic from namespace2/mypod-abcdef"

Why? If NetworkPolicies are dropping traffic it probably means my app is probably broken in some way, but that might be non-obvious. May help with #613 

I think it should work in a similar way to pod events, eg. when waiting for PVC's to become available - it'll only log a message every ~minute and the Kubernetes will fold these into a single message and say it's been repeated n times.

Reply:
Hey @jackevans43 Thanks for the review! I completely agree with you on this, however I think it may be out of the scope of the enhancement since it would require quite a bit more infrastructure to resolve the IPs to pod names and ACLs directly to Network Policies, Etc. However, I fully expect there to be follow up epics regarding network policy observability based on the overall observability epic spearheaded by @russellb

Reply:
I totally agree with the improvement, and that it'd be a future step.  I think it would be good to document this somewhere as one of the next steps with this feature.  Maybe we could have a "future work" section if there's not already something like that in the template.  I would also add integration with the OpenShift console in that "future work" section.

Reply:
We should try and generate an IP audit log in ovn-kube master, since we control IPAM there.

Reply:
That's a good idea.  It's probably already in the northd log, but of course that's not persisted (unless you use cluster logging)